lib/ace/edit_session/bracket_match.js

```diff
@@ -37,17 +37,76 @@ var Range = require("../range").Range;
 
 function BracketMatch() {
 
+
+    this.findSurroundingBrackets = function(position) {
+        if ((position.column == 0) && (position.row)) return null;
+
+        var allBrackets = { 
+            left : [this.$findOpeningBracket("}", position, /(\.?.paren)+/), 
+                    this.$findOpeningBracket(']', position, /(\.?.paren)+/), 
+                    this.$findOpeningBracket(')', position, /(\.?.paren)+/)],
+            right : [this.$findClosingBracket("{", position, /(\.?.paren)+/),
+                     this.$findClosingBracket('[', position, /(\.?.paren)+/),
+                     this.$findClosingBracket('(', position, /(\.?.paren)+/)]
+        };
+
+        var left = {row: null, column: null};
+        var right = {row: null, column: null};
+
+        var leftRowArr = [],
+            leftColArr = [],
+            rightColArr = [],
+            rightRowArr = [];
+
+        var length = Math.max(allBrackets.left.length, allBrackets.right.length)
+
+        for (var key = 0; key < length; key++) {
+            if (allBrackets.left[key] && allBrackets.right[key]) { 
+                leftRowArr.push(allBrackets.left[key].row);
+                rightRowArr.push(allBrackets.right[key].row);
+            }
+        }
+
+        if ((leftRowArr.length !== 0) && (rightRowArr.length !== 0)) {
+            left.row = Math.max.apply(null, leftRowArr);
+            right.row = Math.min.apply(null, rightRowArr);
+        } else {
+            return null;
+        }   
+
+       for (var key = 0; key < length; key++) {
```

Comment:
isn't it easier to introduce two variables in the very first loop at lines 63-68 for the upper bound from the left side and lower bound from the right side, and when the loop is done, check if the result bounds correspond to matching brackets? If brackets match then you're done, otherwise we have bracket mismatch and will want to report it.